source/loader/layers/sanitizer/asan_interceptor.cpp

```diff
@@ -0,0 +1,594 @@
+}
+
+ur_result_t SanitizerInterceptor::releaseMemory(ur_context_handle_t Context,
+                                                void *Ptr) {
+    auto ContextInfo = getContextInfo(Context);
+
+    std::shared_lock<ur_shared_mutex> Guard(ContextInfo->Mutex);
+
+    auto Addr = (uptr)Ptr;
+    // Find the last element is not greater than key
+    auto AllocInfoIt = ContextInfo->AllocatedUSMMap.upper_bound((uptr)Addr);
+    if (AllocInfoIt == ContextInfo->AllocatedUSMMap.begin()) {
+        context.logger.error(
+            "Can't find release pointer({}) in AllocatedAddressesMap", Ptr);
+        return UR_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+    --AllocInfoIt;
+    auto &AllocInfo = AllocInfoIt->second;
+
+    context.logger.debug("USMAllocInfo(AllocBegin={}, UserBegin={})",
+                         AllocInfo->AllocBegin, AllocInfo->UserBegin);
+
+    if (Addr != AllocInfo->UserBegin) {
+        context.logger.error("Releasing pointer({}) is not match to {}", Ptr,
+                             AllocInfo->UserBegin);
+        return UR_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    // TODO: Update shadow memory
+    return context.urDdiTable.USM.pfnFree(Context,
+                                          (void *)AllocInfo->AllocBegin);
+}
+
+bool SanitizerInterceptor::preLaunchKernel(ur_kernel_handle_t Kernel,
+                                           ur_queue_handle_t Queue,
+                                           ur_event_handle_t &Event) {
+    prepareLaunch(Queue, Kernel);
+
+    UR_CALL(updateShadowMemory(Queue));
+
+    // Return LastEvent in QueueInfo
+    auto Context = getContext(Queue);
+    auto ContextInfo = getContextInfo(Context);
+    auto QueueInfo = ContextInfo->getQueueInfo(Queue);
+
+    std::scoped_lock<ur_mutex> Guard(QueueInfo->Mutex);
+    Event = QueueInfo->LastEvent;
+    QueueInfo->LastEvent = nullptr;
+
+    return true;
+}
+
+void SanitizerInterceptor::postLaunchKernel(ur_kernel_handle_t Kernel,
+                                            ur_queue_handle_t Queue,
+                                            ur_event_handle_t *Event) {
+    auto Program = getProgram(Kernel);
+    ur_event_handle_t ReadEvent{};
+
+    // If kernel has defined SPIR_DeviceSanitizerReportMem, then we try to read it
+    // to host, but it's okay that it isn't defined
+    auto Result = context.urDdiTable.Enqueue.pfnDeviceGlobalVariableRead(
+        Queue, Program, kSPIR_DeviceSanitizerReportMem, true,
+        sizeof(SPIR_DeviceSanitizerReportMem), 0,
+        &SPIR_DeviceSanitizerReportMem, 1, Event, &ReadEvent);
+
+    if (Result == UR_RESULT_SUCCESS) {
+        *Event = ReadEvent;
+
+        auto AH = &SPIR_DeviceSanitizerReportMem;
+        if (!AH->Flag) {
+            return;
+        }
+
+        const char *File = AH->File[0] ? AH->File : "<unknown file>";
+        const char *Func = AH->Func[0] ? AH->Func : "<unknown func>";
+
+        fprintf(stderr, "\n====ERROR: DeviceSanitizer: %s on %s\n\n",
```

Comment:
this is now causing warnings on Windows. Might be simpler to add something like "print always" function:
```patch
index 0575660f..27fcca42 100644
--- a/source/common/logger/ur_logger.hpp
+++ b/source/common/logger/ur_logger.hpp
@@ -43,6 +43,11 @@ inline void error(const char *format, Args &&...args) {
     get_logger().log(logger::Level::ERR, format, std::forward<Args>(args)...);
 }
 
+template <typename... Args>
+inline void always(const char *format, Args &&...args) {
+    get_logger().always(format, std::forward<Args>(args)...);
+}
+
 inline void setLevel(logger::Level level) { get_logger().setLevel(level); }
 
 inline void setFlushLevel(logger::Level level) {
diff --git a/source/common/logger/ur_logger_details.hpp b/source/common/logger/ur_logger_details.hpp
index 4759a2fd..bf63f7ff 100644
--- a/source/common/logger/ur_logger_details.hpp
+++ b/source/common/logger/ur_logger_details.hpp
@@ -51,6 +51,12 @@ class Logger {
         log(logger::Level::ERR, format, std::forward<Args>(args)...);
     }
 
+    template <typename... Args> void always(const char *format, Args &&...args) {
+        if (sink) {
+            sink->log(level, format, std::forward<Args>(args)...);
+        }
+    }
+
     template <typename... Args>
     void log(logger::Level level, const char *format, Args &&...args) {
         if (level < this->level) {
```

Reply:
Thanks! Done.

Reply:
I found that this way is not proper as it will add some unwanted prefixes. @pbalcer

Reply:
I didn't test my suggestion :/ Sorry about that. I think I know the problem, see above.

If you don't want the prefix, you can set `skip_prefix` when creating the logger:
`logger::create_logger("sanitizer", true)`

Reply:
> I didn't test my suggestion :/ Sorry about that. I think I know the problem, see above.
> 
> If you don't want the prefix, you can set `skip_prefix` when creating the logger: `logger::create_logger("sanitizer", true)`

Thanks! Done.

Reply:
> If you don't want the prefix, you can set `skip_prefix` when creating the logger:
> `logger::create_logger("sanitizer", true)`

Hi @pbalcer, I changed the logic of the logger so that, when the logger level is QUIET, the prefix will not be printed. 

This way is suitable for my needs, and I guess this is also consistent with what QUIET means?